var/spack/repos/builtin/packages/papi/package.py

```diff
@@ -30,11 +31,59 @@ class Papi(Package):
     version('5.4.1', '9134a99219c79767a11463a76b0b01a2')
     version('5.3.0', '367961dd0ab426e5ae367c2713924ffb')
 
+    variant('example', default=True, description='Install the example files')
+    variant('cuda', default=False, description='Enable CUDA support')
+    variant('nvml', default=False, description='Enable NVML support')
+    variant('infiniband', default=False, description='Enable Infiniband support')
+    variant('powercap', default=False, description='Enable powercap interface support')
+    variant('rapl', default=False, description='Enable RAPL support')
+    variant('lmsensors', default=False, description='Enable lm_sensors support')
+
+    conflicts('+cuda', when='@:5.6.0')
+    conflicts('+nvml', when='@:5.6.0')
+
+    depends_on('cuda', when='+cuda')
+    depends_on('cuda', when='+nvml')
+    depends_on('lm-sensors', when='+lmsensors')
+
     # Does not build with newer versions of gcc, see
     # https://bitbucket.org/icl/papi/issues/46/cannot-compile-on-arch-linux
     patch('https://bitbucket.org/icl/papi/commits/53de184a162b8a7edff48fed01a15980664e15b1/raw', sha256='64c57b3ad4026255238cc495df6abfacc41de391a0af497c27d0ac819444a1f8', when='@5.4.0:5.6.99%gcc@8:')
 
+    def setup_environment(self, spack_env, run_env):
+        if '^cuda' in self.spec:
+            spack_env.set('CUDA_DIR', self.spec['cuda'].prefix)
+            run_env.prepend_path('LD_LIBRARY_PATH',
+                                 self.spec['cuda'].prefix.extras.CUPTI.lib64)
+
+            for path in self.spec['cuda'].libs.directories + self.rpath:
+                if path.find('/stubs') > -1:
+                    spack_env.remove_path('SPACK_RPATH_DIRS', path)
+                    run_env.remove_path('SPACK_RPATH_DIRS', path)
+
+            for path in self.spec['cuda'].libs.directories + \
+                    [self.spec['cuda'].prefix.extras.CUPTI.lib64]:
+                spack_env.append_path('SPACK_LINK_DIRS', path)
```

Comment:
`.libs.directories` for any package will automatically be added to `SPACK_LINK_DIRS` so that should not be required. 

Regarding the addition of `self.spec['cuda'].prefix.extras.CUPTI.lib64`, a couple of questions:

* Do you think all dependents would use these libraries?
* Do all CUDA builds come with CUPTI?

Depending on the answers I may have additional suggestions (e.g. modify `.libs` for the CUDA package, add a variant to the CUDA package, or in fact the way you do it here may be the simplest way for now).

Reply:
CUPTI is an "extra" because it is mostly used by tool builders and low-level tinkerers that want to measure a bit more of hardware performance than the common user. I don't think CUPTI is a guarantee, but it's been part of CUDA for years and all tools for CUDA use it. One good example of Spack package that uses it is `hpctoolkit`. `cbtf-argonavis` does some string operations to add CUPTI in addition to `.prefix.extras` attribute.

Reply:
Assuming CUPTI doesn't have libraries with equivalent names I think it is also appropriate to provide them in the Spack CUDA package. If you know the names of the libraries provided by CUPTI that would be as simple as updating the `find_libraries` call in the `cuda` package to be a list: `find_libraries(['libcuda', ...], ...)`

Reply:
CUPTI only provides `libcupti.so` (as a link to the actual version) in CUDA 8, 9, and 10. If you look into `extras`, you get a debugger and demos. I don't think any packages use it with NVIDIA providing their own debugging facilities.